include/boost/beast/websocket/detail/mask.hpp

```diff
@@ -83,181 +84,59 @@ using maskgen = maskgen_t<std::minstd_rand>;
 
 inline
 void
-prepare_key(std::uint32_t& prepared, std::uint32_t key)
+prepare_key(prepared_key& prepared, std::uint32_t key)
 {
-    prepared = key;
+    prepared[0] = (key >>  0) & 0xff;
+    prepared[1] = (key >>  8) & 0xff;
+    prepared[2] = (key >> 16) & 0xff;
+    prepared[3] = (key >> 24) & 0xff;
 }
 
-inline
+template<std::size_t N>
 void
-prepare_key(std::uint64_t& prepared, std::uint32_t key)
-{
-    prepared =
-        (static_cast<std::uint64_t>(key) << 32) | key;
-}
-
-template<class T>
-inline
-typename std::enable_if<std::is_integral<T>::value, T>::type
-ror(T t, unsigned n = 1)
+rol(std::array<unsigned char, N>& v, unsigned n)
 {
-    auto constexpr bits =
-        static_cast<unsigned>(
-            sizeof(T) * CHAR_BIT);
-    n &= bits-1;
-    return static_cast<T>((t << (bits - n)) | (
-        static_cast<typename std::make_unsigned<T>::type>(t) >> n));
+    auto v0 = v;
+    for(std::size_t i = 0; i < v.size(); ++i )
+        v[i] = v0[(i + n) % v.size()];
 }
 
-// 32-bit optimized
-//
-template<class = void>
-void
-mask_inplace_fast(
-    boost::asio::mutable_buffer const& b,
-        std::uint32_t& key)
-{
-    auto n = b.size();
-    auto p = reinterpret_cast<std::uint8_t*>(b.data());
-    if(n >= sizeof(key))
-    {
-        // Bring p to 4-byte alignment
-        auto const i = reinterpret_cast<
-            std::uintptr_t>(p) & (sizeof(key)-1);
-        switch(i)
-        {
-        case 1: p[2] ^= static_cast<std::uint8_t>(key >> 16); BOOST_FALLTHROUGH;
-        case 2: p[1] ^= static_cast<std::uint8_t>(key >> 8);  BOOST_FALLTHROUGH;
-        case 3: p[0] ^= static_cast<std::uint8_t>(key);
-        {
-            auto const d = static_cast<unsigned>(sizeof(key) - i);
-            key = ror(key, 8*d);
-            n -= d;
-            p += d;
-            BOOST_FALLTHROUGH;
-        }
-        default:
-            break;
-        }
-    }
-
-    // Mask 4 bytes at a time
-    for(auto i = n / sizeof(key); i; --i)
-    {
-        *reinterpret_cast<
-            std::uint32_t*>(p) ^= key;
-        p += sizeof(key);
-    }
-
-    // Leftovers
-    n &= sizeof(key)-1;
-    switch(n)
-    {
-    case 3: p[2] ^= static_cast<std::uint8_t>(key >> 16); BOOST_FALLTHROUGH;
-    case 2: p[1] ^= static_cast<std::uint8_t>(key >> 8);  BOOST_FALLTHROUGH;
-    case 1: p[0] ^= static_cast<std::uint8_t>(key);
-        key = ror(key, static_cast<unsigned>(8*n));
-        BOOST_FALLTHROUGH;
-    default:
-        break;
-    }
-}
-
-// 64-bit optimized
+// Apply mask in place
 //
-template<class = void>
+inline
 void
-mask_inplace_fast(
-    boost::asio::mutable_buffer const& b,
-        std::uint64_t& key)
+mask_inplace(boost::asio::mutable_buffer& b, prepared_key& key)
 {
     auto n = b.size();
-    auto p = reinterpret_cast<std::uint8_t*>(b.data());
-    if(n >= sizeof(key))
+    auto mask = key; // avoid aliasing
+    auto p = reinterpret_cast<unsigned char*>(b.data());
+    while(n >= 4)
```

Comment:
I suspect that using a `sizeof()` to find the best natural type in stead of `4` might generate optimal code on more architectures and compilers.

Reply:
Inconsequential, the compilers know exactly what is going on here in fact on clang (I think) it unrolls the loop and does 16 bytes at a time.

Reply:
Ack